run apps as non-root user (diagexec setresuid)

alban opened the pull request:
Fix #539 by adding `setresuid` and `setresgid` in `diagexec`.

I preferred a solution based on RootDirectory (#679) but this was not ready yet.

alban commented:
Warning about the functional test in this PR: gexpect has a bug with some timing (https://github.com/ThomasRooney/gexpect/pull/5) and I work around the gexpect bug with a sleep.

iaguis commented on stage1/rootfs/diagexec/diagexec.c:
```diff
@@ -180,19 +181,39 @@ static void load_env(const char *env)
-	exe = argv[4];
-
-	load_env(env);
+	uid_str = argv[4];
```
What if the user we get is in text format?

iaguis commented on stage1/rootfs/diagexec/diagexec.c:
```diff
@@ -180,19 +181,39 @@ static void load_env(const char *env)
-
-	load_env(env);
+	uid_str = argv[4];
+	uid = atoi(uid_str);
```
What if the user we get is in text format?

alban replied:
I don't handle that. I'm not sure what we should do in that case...
See also https://github.com/appc/spec/issues/231

alban commented on tests/inspect/inspect.go:
```diff
@@ -112,13 +114,21 @@ func main() {
+	if globalFlags.PrintUser {
+		// gexpect is racy: workaround it here
+		// TODO: fix it in gexpect, see: https://github.com/ThomasRooney/gexpect/pull/5
+		time.Sleep(10 * time.Millisecond)
```
The fix https://github.com/ThomasRooney/gexpect/pull/6 is merged.
I will update this...

alban commented:
Branch updated. It contains the fix for gexpect and the `sleep` has been removed.

vcaputo commented:
Does it look like this will all be obsoleted by using a version of systemd newer than 215?

alban commented:
github hid some comments after I rebased this branch.

@vcaputo @iaguis : I didn't want to support for all forms of user/group but only numeric uid/gid for now. I added a check with error reporting. The only non-numeric user/group I accept is the string "root" because it was working before so I don't want to introduce the regression. (sytemd has special-case code for "root")

It is a stop-gap and should be revisited either when we get a better idea from https://github.com/appc/spec/issues/231 or when we let systemd do the setuid instead of diagexec. 

This stop-gap would stay even when using systemd v219 because support for `ExecFail=/usr/bin/diagexec` is not implemented yet.

I reduced the verbosity in diagexec by using the first code snippet from your review. `setresuid` and `setresgid` are optional if we want to run as root and we are already root.

alban commented:
I'm not sure why Travis failed to test this.

Semaphore passed the tests successfully on this branch (commit 76e6ea6). I don't know why I don't see the [link to the Semaphore results](https://semaphoreci.com/coreos/rkt/branches/pull-request-805/builds/6).

krnowak commented:
I also had some bogus failure in #799. I suppose that travis was simply overloaded.

vcaputo commented on stage1/rootfs/diagexec/diagexec.c:
```diff
@@ -180,19 +181,29 @@ static void load_env(const char *env)
-	pexit_if(chroot(root) == -1, "Chroot \"%s\" failed", root);
-	pexit_if(chdir(cwd) == -1, "Chdir \"%s\" failed", cwd);
+	pexit_if(*root && chroot(root) == -1, "Chroot \"%s\" failed", root);
+	pexit_if(*cwd && chdir(cwd) == -1, "Chdir \"%s\" failed", cwd);
```
Is it expected that either cwd or root would be "" and should be silently ignored if so?

It seems incorrect for diagexec to have a **successful** code path which potentially doesn't attempt to chroot or chdir.

alban commented:
Thanks for the review @vcaputo. I updated the branch.

vcaputo commented:
ok cool, lgtm

jonboulle commented:
awwwwwwwwwwwwwwww yeah

jonboulle commented:
@alban what is the latest on ExecFail?

alban commented:
@jonboulle I have not tried to implement ExecFail in systemd yet...
The bug is open on:
https://bugs.freedesktop.org/show_bug.cgi?id=89840

jonboulle commented:
@alban something that came up in OOB discussion, it's unclear how ExecFail would actually work w.r.t. the chroot - is it going to inject a diagnostic exec into the chroot itself? or expected do the chroot itself? which would make it inconsistent with the other Exec* parameters in that case...
